src/directives/if-defined.ts

```diff
@@ -14,19 +14,27 @@
 
 import {AttributePart, directive, Part} from '../lit-html.js';
 
+const previousValues = new WeakMap<Part, unknown>();
+
 /**
  * For AttributeParts, sets the attribute if the value is defined and removes
  * the attribute if the value is undefined.
  *
  * For other part types, this directive is a no-op.
  */
 export const ifDefined = directive((value: unknown) => (part: Part) => {
+  const previousValue = previousValues.get(part);
+
   if (value === undefined && part instanceof AttributePart) {
-    if (value !== part.value) {
+    // If the value is undefined, remove the attribute, but only if the value
+    // was previously defined.
+    if (value !== previousValue || !previousValues.has(part)) {
```

Comment:
Nit: we already know `value === undefined`, so I think it's clearer if we just test `previousValue` against `undefined`. And when defining it that way, we can reason about the `!previousValues.has(part)`. If `previousValues` doesn't contain `part`, that implies `previousValue === undefined`. So our check is actually:

```js
if (previousValue !== undefined || previousValue === undefined) {}
```

Which is always true. So let's just delete the if statement.

Reply:
`!previousValues.has(part)` is different than `previousValue !== undefined` though. A map will return `undefined` for both the case where the value wasn't present, and when it was but the value actually was `undefined`. Here we're checking if this is the first time we've seen this part, and if so removing the attribute.

I added a test that fails without this check.

Reply:
`!previousValues.has(part)` is equivalent to `previousValue === undefined` (note no `!`).

The `previousValue !== undefined` (note `!`) is equivalent to `previousValue !== value`, because we know `value === undefined`.

Reply:
`!previousValues.has(part)` implies `previousValue === undefined`

`previousValue === undefined` does not imply `!previousValues.has(part)`

So we can't replace `!previousValues.has(part)` with `previousValue === undefined` and the expression doesn't reduce.

It's not that big of a deal though, at most it would lead to one extra call to `removeAttribute`.

Reply:
Ohh, I see what you're saying now. If we `previousValues.set(part, undefined)`, we don't want to re-remove the attribute.